### PR TITLE
Implement basic mapping capabilities for YAML.

### PR DESCRIPTION
**Pull request by robertwb**
Examples can be found in the doc.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by robertwb** [comment]
R: @bzablocki

File: sdks/python/apache_beam/yaml/readme_test.py [hunk 1]
```diff
@@ -211,28 +215,39 @@ def parse_test_methods(markdown_lines):
...
-    return type(
-        'ReadMeTest', (unittest.TestCase, ), dict(parse_test_methods(readme)))
+def createTestSuite(name, path):
+  with open(path) as readme:
+    return type(name, (unittest.TestCase, ), dict(parse_test_methods(readme)))
 
 
-ReadMeTest = createTestSuite()
+ReadMeTest = createTestSuite(
+    'ReadMeTest', os.path.join(os.path.dirname(__file__), 'README.md'))
```
[the remark this answers is not in the document]
Comment (robertwb) [reply sdks/python/apache_beam/yaml/readme_test.py:247]: I have such a block in yaml_mapping_test. I'm open to putting it here instead if you think that'd be preferable.

File: sdks/python/apache_beam/yaml/yaml_mapping.py [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+
+    # We should support this for all languages.
+    language = "python"
+
+  if language in ("sql", "calcite"):
+    selects = [f'{expr} AS {name}' for (name, expr) in fields.items()]
+    query = "SELECT " + ", ".join(selects) + " FROM PCOLLECTION"
+    if keep:
+      query += " WHERE " + keep
+    print("QUERY", query)
```
[the remark this answers is not in the document]
Comment (robertwb) [reply sdks/python/apache_beam/yaml/yaml_mapping.py:196]: Leftover debugging. Removed.

File: sdks/python/apache_beam/yaml/yaml_mapping.py [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+            'cross_product': cross_product,
+        },
+        **language_keywords
+    })
+
+  else:
+    # TODO(yaml): Support javascript expressions and UDFs.
+    # TODO(yaml): Support java by fully qualified name.
+    # TODO(yaml): Maybe support java lambdas?
+    raise ValueError(f'Unknown language: {language}')
```
[the remark this answers is not in the document]
Comment (robertwb) [reply sdks/python/apache_beam/yaml/yaml_mapping.py:225]: Done.

File: sdks/python/apache_beam/yaml/yaml_mapping_test.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+              beam.Row(a=3, b='y', c=.125, range=2),
+          ]))
+
+
+YamlMappingDocTest = createTestSuite(
+    'YamlMappingDocTest',
+    os.path.join(os.path.dirname(__file__), 'yaml_mapping.md'))
+
+if __name__ == '__main__':
+  logging.getLogger().setLevel(logging.INFO)
```
[the remark this answers is not in the document]
Comment (robertwb) [reply sdks/python/apache_beam/yaml/yaml_mapping_test.py:136]: Fixed.

**Review by robertwb — COMMENTED** [review]
Thanks.

File: sdks/python/apache_beam/yaml/yaml_provider.py [hunk 5]
```diff
@@ -59,7 +60,11 @@ def provided_transforms(self) -> Iterable[str]:
     raise NotImplementedError(type(self))
 
   def create_transform(
-      self, typ: str, args: Mapping[str, Any]) -> beam.PTransform:
+      self,
+      typ: str,
+      args: Mapping[str, Any],
+      yaml_create_transform: Callable[[Mapping[str, Any]], beam.PTransform]
```
Comment (robertwb) [inline sdks/python/apache_beam/yaml/yaml_provider.py:66]: FYI, I'm looking at refactoring this a bit (doing a re-write like how the other "meta" transforms are handled, as this is a bit hacky.

Comment (robertwb) [reply sdks/python/apache_beam/yaml/yaml_provider.py:66]: This turns out to have its own non-trivial complications. Keeping as is.

**Review by bzablocki — APPROVED** [review]

**Comment by robertwb** [comment]
10:12:33 ERROR: Step ‘Publish JUnit test result report’ failed: No test report files were found. Configuration error? errors look unrelated, and these all passed before fixing the trivial lint issues.

**Comment by Abacn** [comment]
The error message appears all Jenkins jobs, now they are not test anything.

**Comment by Abacn** [comment]
PythonLint indeed breaks after this PR:

```
apache_beam/yaml/readme_test.py:229:42: E1133: Non-iterable value code_lines is used in an iterating context (not-an-iterable)
```

CC: @robertwb @bzablocki

**Pull request merged**